### PR TITLE
Model lp returns

### PR DESCRIPTION
**Pull request by jepidoptera**

### CODE REVIEW

File: hydradx/notebooks/Omnipool/LP_returns_analysis.ipynb [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "price_path = []\n",
+    "modifier_list = [-0.25, 0, 0.25, 0.5, 0.75]\n",
+    "colors = pu.color_gradient(length=5, color1=(255, 0, 0), color2 = (0, 255, 0))\n",
+    "initial_price_sum = sum([price_list[0][tkn] for tkn in price_list[0]])\n",
```
Comment (poliwop) [inline hydradx/notebooks/Omnipool/LP_returns_analysis.ipynb:143]: It is not meaningful to sum prices in this way. Prices are in units USD/BTC, USD/ETH, USD/DOT, etc, and summing them is not correct. Whatever is the highest price (BTC) will dominate the sum, even if BTC makes up a tiny portion of the Omnipool. Then the aggregate graph will be something close to the BTC graph but won't reflect the overall dynamics of the Omnipool.

The quickest fix is to just do this graph with BTC price since that's approximately what it is currently. We could also do the other assets individually, and graph the result for each asset.

If you want some aggregate measure, you can do something like take % change of each asset and then do a weighted average. But I think it's better to just look at each asset individually.

File: hydradx/notebooks/Omnipool/LP_returns_analysis.ipynb [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+    "price_events = []\n",
+    "for i, scenario in enumerate(price_path):\n",
+    "    initial_state._evolve_function = historical_prices(scenario)\n",
+    "    random.seed(42)\n",
+    "    events = run.run(initial_state, time_steps)\n",
+    "    price_events.append(events[::100])\n",
+    "    returns.append([cash_out_omnipool(event.pools['omnipool'], event.agents['LP'], event.external_market) for event in price_events[i]])\n",
+    "    dot_price.append(pu.get_datastream(price_events[i], asset='DOT'))\n",
+    "    \n",
+    "print('Trade volume per day as a fraction of TVL:', \n",
```
[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/notebooks/Omnipool/LP_returns_analysis.ipynb:204]: This can be a follow-up task, doesn't need to happen in this PR.

File: hydradx/notebooks/Omnipool/LP_returns_analysis.ipynb [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
...9xh7fgDhHgWCIh8dHYcfLAxF6rkkUeAbdtESHCTJmKZ+j0aP160QfLLL0KcltVyVebMCV2WLFWDCYLYPzBfu486SjxDm6/xdR0p1s2caXQ7Z2YC+fnWAz/Vdda98w7w/vvGPHlK4TmCSBaq46zrAuBPznkJ57wKwI8ARgDoCEA+lX8D4FRt+iQA73DOyznnqwAshxD6BgBYzjlfyTmvAPCOti1BEHUMKaxxzoPTfrM5JRCwruyk4VrrkAzcCjUfkZ0TRi0R7/cbO2iROoXiWVXw5ptFp/OQQ+J3TIKoCcxV6KIRJDp0AK69NrJtzZWXAWDqVHA7B5WVI0o+dMsiA9UlnFhn5TxMBLKi9qRJohjGmjVi/uqrRefjzz+t97Or9huLWKf+P6Q7wup/qiKddevWGd18Dz4IHHEEsGuXXjxDbduECcCMGWK+TRt93ZNPitf27YH77gs939y5oiCHpKzM2DEcN85YnESGV5sdmrt2AW++GXp8WR2ZIAgiHpgHcyMZkK6LyHvFtdcCzzwjppctA15+WRQJsiJeYbDxLFBBEDVAdcS6+QAGM8YaMcbSAQwHUABgAXSxbaS2DABaAChS9l+nLbNbThBEXcJUQbBK6+f6zf1dzp3Fumhztr//vkhY/9xz9tuoydX9fj1k6+KLRfgcQRDWjB8P/PWXKAgxbJhlvkdHzIUOouGee8Bfn2a9zil80U6EipZwYp1dIYbaxs4NcP/9QNeu9vnq7Dp+sbwvVfSKRqyrqgIKCqxdkA8+GBoWtWkTcM01+vzzzwP//CMKO8jzZWYaB2EmTNCn//1Xn542zfp7dOut4lV+Pm+8YVwvxVAzJNYRBBFPzGJdJMV76iLqveKHH8TrzTc771OdMFj1um+Vm5cgkoiYxTrO+SIA4wF8DWAGgNkA/AAuAnAFY2wWgCwAcfsVMMbGMMb+YYz9s2XLlngdliCIeKA6DRgzFJkwkJcX2olTwqXMYa+2x5E0bw507w6MHWu/jfqA4/eLMCtAhN6pzgyCIIx4vcABB4hcLl9+GX1nIZIHYQe3AC+2cdo6iXXx6tCEO06sxQniTaNG1svDuSD37LFeHo2zjnPxp/4/5ABIuOO43bor0CpMescO0RHr0cP+s05LExVXH3hAz3OYmWkUzq680npfn8/6e/TII8Crr+rn3G6qUC6dH4BwSMuQ3UMPtT4PQ
```
Comment (poliwop) [inline hydradx/notebooks/Omnipool/LP_returns_analysis.ipynb:308]: Where do we have data supporting that we have observed 5% 24hr trade volume recently? I think it's lower. We should remove this claim or have data to support it IMO.

Comment (poliwop) [reply hydradx/notebooks/Omnipool/LP_returns_analysis.ipynb:308]: Also, can we put legends on these graphs so we can tell which color corresponds to what value?

File: hydradx/model/amm/omnipool_amm.py [hunk 4]
```diff
@@ -286,7 +286,8 @@ def __init__(self, state: OmnipoolState):
         # self.sub_pools = copy.deepcopy(state.sub_pools)
         self.oracles = {k: OracleArchiveState(v) for (k, v) in state.oracles.items()}
         self.unique_id = state.unique_id
-
+        self.volume_in = state.current_block.volume_in
```
Comment (poliwop) [inline hydradx/model/amm/omnipool_amm.py:289]: Are these changes an unrelated bug fix?

Comment (jepidoptera) [reply hydradx/model/amm/omnipool_amm.py:289]: No, this is so I can track trade volume as a percentage of the pool, because I think that's relevant to the returns LPs will get

File: hydradx/notebooks/Omnipool/LP_returns_analysis2.ipynb [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+    "    },\n",
+    "    agents = {\n",
+    "        'Trader': Agent(\n",
+    "            holdings={tkn: 10000000 for tkn in assets},\n",
+    "            trade_strategy=omnipool_arbitrage(\n",
+    "                pool_id='omnipool'\n",
+    "            )\n",
+    "        ),\n",
+    "        'LP': Agent(\n",
+    "            holdings={'INTR': 10000 / price_list[0]['DOT']},\n",
```
Comment (poliwop) [inline hydradx/notebooks/Omnipool/LP_returns_analysis2.ipynb:77]: Is initializing based on the price of DOT intentional?

File: hydradx/notebooks/Omnipool/LP_returns_analysis2.ipynb [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+    "colors = pu.color_gradient(length=5, color1=(255, 0, 0), color2 = (0, 255, 0))\n",
+    "initial_dot_price = price_list[0]['DOT']\n",
+    "\n",
+    "plt.figure(figsize=(20, 5))\n",
+    "plt.title(\"Price paths from optimistic to pessimistic\")\n",
+    "\n",
+    "for i, price_shift in enumerate(modifier_list):\n",
+    "    price_shift_per_block = price_shift / time_steps\n",
+    "    price_path.append([\n",
+    "        {('INTR' if tkn == 'DOT' else tkn): (\n",
```
Comment (poliwop) [inline hydradx/notebooks/Omnipool/LP_returns_analysis2.ipynb:123]: I don't understand what happens do the DOT price here? Are we somehow starting with DOT in the Omnipool and not INTR and then switching?

Comment (poliwop) [reply hydradx/notebooks/Omnipool/LP_returns_analysis2.ipynb:123]: To clarify.... isn't this going to result in the key 'INTR' getting written to twice in this dict (and I think overwritten the 2nd time) but the key 'DOT' getting written to 0 times?

File: hydradx/notebooks/Omnipool/LP_returns_analysis2.ipynb [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "price_path = []\n",
+    "modifier_list = [-0.5, -0.25, 0, 0.25, 0.5]\n",
+    "colors = pu.color_gradient(length=5, color1=(255, 0, 0), color2 = (0, 255, 0))\n",
+    "initial_dot_price = price_list[0]['DOT']\n",
+    "\n",
+    "plt.figure(figsize=(20, 5))\n",
```
Comment (poliwop) [inline hydradx/notebooks/Omnipool/LP_returns_analysis2.ipynb:117]: This plot doesn't graph anything for me, just empty axes.

File: hydradx/notebooks/Omnipool/LP_returns_analysis2.ipynb [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+   "source": [
+    "returns = []\n",
+    "intr_prices = []\n",
+    "price_events = []\n",
+    "plt.figure(figsize=(20, 5))\n",
+    "for i, scenario in enumerate(price_path):\n",
+    "    initial_state._evolve_function = historical_prices(scenario)\n",
+    "    random.seed(42)\n",
+    "    events = run.run(initial_state, time_steps)\n",
+    "    price_events.append(events[::200])\n",
```
Comment (poliwop) [inline hydradx/notebooks/Omnipool/LP_returns_analysis2.ipynb:188]: Am I understanding correctly that we're only graphing every 200th data point?

File: hydradx/notebooks/Omnipool/LP_returns_analysis2.ipynb [hunk 9]
```diff
@@ -0,0 +1,437 @@
...
...57/UjWEwFo5nY7DjiwO22Lbf+8eFbHy5e21mkImjAdp/Ao/D/6CB3x9IkMhoAYqaTllSypXPa5oFtpJsEXoHKkPK8RBqdrgbe9Px5nYvGLhaaHrcqj/72HRgYDeuUQ4sYMISshEmNeLv4Xsjh5fgt28XFc/G4mztXqEdisBwGvz7Fsu369VDdRKBQKheJWw9kZ4kp/InyjEjzyFdxb3IExKxOP1ncS/u4okrasBDs73Jt3KHa7hPWLSVi/GIBa7y3Ds213m/V2ru602p5N5tnjpa4AViQXL8LevdCrV/nHKCuhoRKdPHs2tG599farUCgUBdDs7Ql46DlifpqNnpuDZu9AxulQkndswDGoGqdfGGDuG792EfFrF5G6byvxaxfZjNNo8QFc6jQlZecmszAS99s3uDfvwNm3n8TBP5hqL3+BY0AIrg1aFjufiPfHE/3jRzZtVSd8gKZpGBxFMDGmp3BibB8qDZ1Q6H/DidE9Sdn1J3U+Wo1Xx3vK9Froubmk7v8H5xoNiFn2GdlRZ6n87Ds4+AcTPu1pUvdtxb1lZ2J//pzKz75D8IiXzNuOGzeOGTNmkJKSUuTYe/bsISIigj59+jBz5kxze2pqKu+99x4bNmy4btO8QAk/V5zH97cknPBC7QfSbPMaG4VCtiOcqQF+cRB0EVruAcdJQ6FKFZgxAy08HJdt26j+3jSCDx/BJROMGlysBG7pkGeQtC7PZDAYRfypdFHSxJyzAGIBqHEWqbiVmQkffgiJiTB0KNq58/hPnmkzr1quEN0kkOCPFuM84xMwiT6//CIlbr29K/gVUygUCoXiBsHJCXbuhDFjIDkZ+vaVtOW9e+Hpp6FevWs9wwrDo03XQm0utZticHYt1O5//0hSD24n89Qhc5tXxz5FjqvZO+BS+zLukGZnQ1CQPI+IkHOmK01CgqWgRZs28NVXUtV09GhJB1QoFIqrjFvjtsQgnmqpB7YRPu3pQn3qzllH8n8bufjtDOLXLsLeLwiPlp1I3rGBvJREIr9+h5pvfUfkF28AUG3yZ4RPf4azbz8JSOrXqfGS2lvjrUX4dB+EIT+Cx4Su60QvketFp+r1CRo2Ef/7LRE0mpMIPym7/iR5+3qS/1mDX78nqD7lMzR7B3KTE0jZ9ScAJ8f1waNtD6q/8iUOAZWLLAGf+PcqTo2/F79+T1DthU848+owEv
```
Comment (poliwop) [inline hydradx/notebooks/Omnipool/LP_returns_analysis2.ipynb:223]: *Interlay* price paths

Comment (poliwop) [inline hydradx/notebooks/Omnipool/LP_returns_analysis2.ipynb:352]: I think this is overestimating trade volume. We don't have to make a claim on how much we have seen though, I think we can just provide the 0.3% to 10% range without further comment, they can look on our app to see what the last 24hr trade volume is for each asset.

Comment (poliwop) [inline hydradx/notebooks/Omnipool/LP_returns_analysis2.ipynb:393]: I'm not sure we are seeing clean results here, the baseline results were obtained using our arbitrage agent, while the subsequent results were obtained using an agent that would be different from the arb agent *even if fees were zero*.

File: hydradx/notebooks/Omnipool/LP_returns_analysis2.ipynb (hunk 6 above)
Comment (jepidoptera) [reply hydradx/notebooks/Omnipool/LP_returns_analysis2.ipynb:123]: INTR should only get written to once, since it is not present in the initial price list. I'm just using DOT price as a stand-in for INTR, but if there is data available for INTR I should use that instead.

**Review by poliwop — COMMENTED** [review]
changes left:
- Interlay should be 3% of pool, remaining weight that was INTR should be DOT
- Clarify in text that price paths are INTR price paths
- Move to vault folder

File: hydradx/notebooks/Omnipool/LP_returns_analysis3.ipynb [hunk 10]
```diff
@@ -47,7 +44,8 @@
     "    'USD': {'usd price': 1, 'weight': 0.20},\n",
     "    'BTC': {'usd price': price_list[0]['BTC'], 'weight': 0.10},\n",
     "    'ETH': {'usd price': price_list[0]['ETH'], 'weight': 0.40},\n",
-    "    'INTR': {'usd price': price_list[0]['DOT'], 'weight': 0.20}\n",
+    "    'INTR': {'usd price': price_list[0]['DOT'], 'weight': 0.03},\n",
+    "    'TKN': {'usd price': price_list[0]['DOT'], 'weight': 0.17}\n",
```
Comment (poliwop) [inline hydradx/notebooks/Omnipool/LP_returns_analysis3.ipynb]: This looks incorrect, don't we want TKN to have 3% weight and DOT to have 17% weight, and we want to remove INTR altogether?

File: hydradx/notebooks/Omnipool/LP_returns_analysis3.ipynb [hunk 11]
```diff
@@ -304,18 +308,15 @@
   },
```
Comment (poliwop) [inline hydradx/notebooks/Omnipool/LP_returns_analysis3.ipynb]: This text still mentions INTR

**Review by poliwop — CHANGES_REQUESTED** [review]

**Comment by poliwop** [comment]
Also we still need to move the final notebook to the vault and remove the other two from the PR

File: hydradx/notebooks/Omnipool/LP_returns_analysis.ipynb (hunk 1 above)
Comment (jepidoptera) [reply hydradx/notebooks/Omnipool/LP_returns_analysis.ipynb]: This is resolved, I'm now looking at TKN/LRNA vs USD/LRNA prices.

File: hydradx/notebooks/Omnipool/LP_returns_analysis.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (jepidoptera) [reply hydradx/notebooks/Omnipool/LP_returns_analysis.ipynb]: Resolved because I removed the fee analysis.

File: hydradx/notebooks/Omnipool/LP_returns_analysis3.ipynb (hunk 10 above)
Comment (jepidoptera) [reply hydradx/notebooks/Omnipool/LP_returns_analysis3.ipynb]: You're right, that was incorrect.

**Comment by jepidoptera** [comment]
I see how that somehow the pull_request check failed while the push check succeeded, even though they are the same. All tests pass for me locally, so I guess that's a fluke?

File: hydradx/notebooks/Omnipool/LP_returns_analysis3.ipynb (hunk 11 above)
Comment (poliwop) [reply hydradx/notebooks/Omnipool/LP_returns_analysis3.ipynb]: @jepidoptera this still isn't fixed, but I'm going to approve anyway and make a separate ticket to fix it, because I don't wanna keep going back and forth.

**Review by poliwop — APPROVED** [review]
